Replace Boost.Process with cpp-subprocess

hebasto opened the pull request:
Closes https://github.com/bitcoin/bitcoin/issues/24907.

This PR is based on **theStack**'s [work](https://github.com/bitcoin/bitcoin/issues/24907#issuecomment-1466087049).

The `subprocess.hpp` header has been sourced from the [upstream repo](https://github.com/arun11299/cpp-subprocess) with the only modification being the removal of convenience functions, which are not utilized in our codebase.

Windows-related changes will be addressed in subsequent follow-ups.

theStack committed:
external_signer: replace boost::process with cpp-subprocess

This primarily affects the `RunCommandParseJSON` utility function.

theStack committed:
build: remove boost::process dependency for building external signer support

maflcko reviewed (commented):
I guess the compile error is due to subprocess requiring C++17 and not allowing C++20 (or later)?

fanquake commented on test/lint/lint-spelling.py:
```diff
@@ -12,7 +12,7 @@
 
 IGNORE_WORDS_FILE = 'test/lint/spelling.ignore-words.txt'
-FILES_ARGS = ['git', 'ls-files', '--', ":(exclude)build-aux/m4/", ":(exclude)contrib/seeds/*.txt", ":(exclude)depends/", ":(exclude)doc/release-notes/", ":(exclude)src/leveldb/", ":(exclude)src/crc32c/", ":(exclude)src/qt/locale/", ":(exclude)src/qt/*.qrc", ":(exclude)src/secp256k1/", ":(exclude)src/minisketch/", ":(exclude)contrib/guix/patches"]
+FILES_ARGS = ['git', 'ls-files', '--', ":(exclude)build-aux/m4/", ":(exclude)contrib/seeds/*.txt", ":(exclude)depends/", ":(exclude)doc/release-notes/", ":(exclude)src/leveldb/", ":(exclude)src/crc32c/", ":(exclude)src/qt/locale/", ":(exclude)src/qt/*.qrc", ":(exclude)src/secp256k1/", ":(exclude)src/minisketch/", ":(exclude)src/util/subprocess.hpp", ":(exclude)contrib/guix/patches"]
```
Probably easier to just fix the typos, than add more things for linters to exclude? No new ones should be introduced.

fanquake commented:
Concept ACK on removing Boost Process. Is the plan to also delete any unused code from this header? I don't think copy-pasting dead/untested code into this repository is a good idea, and only serves to complicate review, and potentially cause issues in future (it maybe already doesn't work with c++20?)

hebasto commented:
> Is the plan to also delete any unused code from this header?

I haven't consider it yet.

> it maybe already doesn't work with c++20?

CI jobs that uses C++20 are passing so far.

fanquake commented:
Note that the last version from upstream was > 3 years ago.

hebasto commented:
> I'd rather take the small(er) portion of any code we need, and have something better integrated/reviewed.

Maybe a subtree from our own fork? (to avoid bringing testing stuff to the main repo)

fanquake commented:
> Maybe a subtree from our own fork? 

I think that would mostly just re-introduce the awkwardness we ended up with with univalue, for not much benefit, and for code that shouldn't be critical (i.e compared to leveldb & libsecp256k1), this is for an optional feature, that is not used by a large proportion of the people running this software.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/28981#issuecomment-1836341684

> I'd rather take the small(er) portion of any code we need, and have something better integrated/reviewed. 

Looking at the coverage report of the header, there seems to be some potential for pruning out (89.2% function coverage and 67.4% line coverage),

hebasto commented:
I switched to the `Popen` constructor that accepts `const std::string&` instead of `std::vector<std::string>` to address quoting issues on Windows.

@theStack 
Since you initially chose to use a vector, do you have any objections to this approach?

---

Here are the Windows functional tests -- https://github.com/hebasto/bitcoin/actions/runs/7072075125/job/19250502576

hebasto commented:
CI turned green. Undrafted.

cc @achow101 @Sjors

theStack commented:
> I switched to the `Popen` constructor that accepts `const std::string&` instead of `std::vector<std::string>` to address quoting issues on Windows.
> 
> @theStack Since you initially chose to use a vector, do you have any objections to this approach?

No objections at all. I would have expected that passing the arguments as a single string is _more_ prone to quoting issues than if it's done in a list, where quoting shouldn't be needed at all, but if it works, all the better :) (Maybe that unverified assumption made me use the std::vector `Popen` ctor back then, but I honestly don't really remember).

hebasto commented:
> > I switched to the `Popen` constructor that accepts `const std::string&` instead of `std::vector<std::string>` to address quoting issues on Windows.
> > @theStack Since you initially chose to use a vector, do you have any objections to this approach?
> 
> No objections at all. I would have expected that passing the arguments as a single string is _more_ prone to quoting issues than if it's done in a list, where quoting shouldn't be needed at all, but if it works, all the better :) (Maybe that unverified assumption made me use the std::vector `Popen` ctor back then, but I honestly don't really remember).

To support the string-based approach, let's consider `echo`-ing a JSON-parsable string on Windows:
```
>echo {"success": true}
{"success": true}

```
and reproducing it using Python's subprocess module, which interface is mimicked in cpp-subprocess: 
```py
>>> import subprocess
>>> subprocess.run("cmd.exe /c echo {\"success\": true}")
{"success": true}
CompletedProcess(args='cmd.exe /c echo {"success": true}', returncode=0)
```
It becomes intricate when using a list of strings:
```py
>>> subprocess.run(["cmd.exe", "/c", "echo", "{\"success\": true}"])
"{\"success\": true}"
```
or
```py
>>> subprocess.run(["cmd.exe", "/c", "echo", "{\"success\":", "true}"])
{\"success\": true}
```
or
```py
>>> subprocess.run(['cmd.exe', '/c', 'echo', '{"success":', 'true}'])
{\"success\": true}
```

hebasto commented:
Rebased on top of the #28989.

luke-jr reviewed (changes requested):
>>Maybe a subtree from our own fork?

>I think that would mostly just re-introduce the awkwardness we ended up with with univalue, for not much benefit, and for code that shouldn't be critical (i.e compared to leveldb & libsecp256k1), this is for an optional feature, that is not used by a large proportion of the people running this software.

Which means it should be an entirely external dependency we just check for in configure...

The only thing worse than a subtree, is entirely vendoring it like this PR currently does.

Sjors commented:
The code changes for external signing look pretty minimal, so that's nice. I have no informed opinion on the difference between Boost Process and cpp-process. Except that the latter seems self contained.

> Which means it should be an entirely external dependency we just check for in configure…

In theory that makes sense to me as well, e.g. a fork under the bitcoin-core project, with functionality (mostly) limited to what we need. However in practice it means having to package it and get into distros. Or force anyone compiling Bitcoin to also clone and build this library from source themselves. The latter could be less of a problem, if we make it easier to build _and use_ individual `/depends` packages. But that's not going to be anytime soon.

hebasto commented:
Rebased on top of the merged #28989.

Sjors commented:
Tested on Ubuntu 23.10.

This slightly changes the way strings are escaped, please use this:

```diff
--- a/src/external_signer.cpp
+++ b/src/external_signer.cpp
@@ -60,43 +60,43 @@ bool ExternalSigner::Enumerate(const std::string& command, std::vector<ExternalS
     return true;
 }
 
 UniValue ExternalSigner::DisplayAddress(const std::string& descriptor) const
 {
-    return RunCommandParseJSON(m_command + " --fingerprint " + m_fingerprint + NetworkArg() + " displayaddress --desc \"" + descriptor + "\"");
+    return RunCommandParseJSON(m_command + " --fingerprint " + m_fingerprint + NetworkArg() + " displayaddress --desc " + descriptor);
 }

     const std::string command = m_command + " --stdin --fingerprint " + m_fingerprint + NetworkArg();
-    const std::string stdinStr = "signtx \"" + EncodeBase64(ssTx.str()) + "\"";
+    const std::string stdinStr = "signtx " + EncodeBase64(ssTx.str());
```

That's probably for the better, but it's important to manually test the HWI address verify and PSBT sign feature.

For anyone testing with HWI, consider also testing #24313.

hebasto commented:
Rebased and implemented @Sjors's [suggestions](https://github.com/bitcoin/bitcoin/pull/28981#issuecomment-1854029824).

Thank you @Sjors for testing!

hebasto commented:
> For anyone testing with HWI, consider also testing #24313.

Here is the combined [`231130-replace-bp+pr24313`](https://github.com/hebasto/bitcoin/tree/231130-replace-bp%2Bpr24313) branch.

Sjors commented:
tACK b67b6ab3b69d45034b17f4704a1f3cedb07e9af6

Tested on macOS 14.2.1, Windows 11 Home (using cross compiled Guix build) and Ubuntu 23.10.

I didn't test Windows native. It might be worth splitting off the last commit to get this merged faster. I'm assuming it's more important to get rid of boost process than it is to support native Windows builds.

Commit f6a3b5ef643d4993f0ae3faa0c5ff1ae62ad71d8 which switches the external signer code to use cpp-subprocess looks correct.

I perused the cpp-subprocess implementation. It seems well structured and documented, but I didn't study it in great detail.  

Guix hashes:

```
# soon(tm)
```

theStack reviewed (commented):
Nit: if you have to retouch, could update the commit hash reference in the first commit body (the last upstream patch https://github.com/arun11299/cpp-subprocess/pull/98 was merged a few weeks ago). Regarding the TODO in the same message, are there any more changes included here which would make sense to upstream? Comparing cpp-subprocess master branch with the version included here still shows some small differences:
```diff
--- /home/thestack/bitcoin/cpp-subprocess/subprocess.hpp	Tue Feb 27 14:46:14 2024
+++ ./src/util/subprocess.hpp	Tue Feb 27 14:41:21 2024
@@ -61,7 +61,7 @@
 
 extern "C" {
 #ifdef __USING_WINDOWS__
-  #include <Windows.h>
+  #include <windows.h>
   #include <io.h>
   #include <cwchar>
 
@@ -155,7 +155,7 @@
 //--------------------------------------------------------------------
 
 //Environment Variable types
-#ifndef _MSC_VER
+#ifndef __USING_WINDOWS__
 	using env_string_t = std::string;
 	using env_char_t = char;
 #else
@@ -184,7 +184,7 @@
     //
 
     if (force == false && argument.empty() == false &&
-        argument.find_first_of(L" \t\n\v\"") == argument.npos) {
+        argument.find_first_of(L" \t\n\v") == argument.npos) {
       command_line.append(argument);
     }
     else {
@@ -1057,6 +1057,7 @@
 public:
   Communication(Streams* stream): stream_(stream)
   {}
+  Communication(const Communication&) = default;
   void operator=(const Communication&) = delete;
 public:
   int send(const char* msg, size_t length);
@@ -1094,6 +1095,7 @@
 {
 public:
   Streams():comm_(this) {}
+  Streams(const Streams&) = default;
   void operator=(const Streams&) = delete;
 
 public:
@@ -1426,6 +1428,10 @@
 
 inline int Popen::poll() noexcept(false)
 {
+#ifndef __USING_WINDOWS__
+  if (!child_created_) return -1; // TODO: ??
+#endif
+
 #ifdef __USING_WINDOWS__
   int ret = WaitForSingleObject(process_handle_, 0);
   if (ret != WAIT_OBJECT_0) return -1;
@@ -1439,8 +1445,6 @@
 
   return retcode_;
 #else
-  if (!child_created_) return -1; // TODO: ??
-
   int status;
 
   // Returns zero if child is still running
```

hebasto committed:
Add `cpp-subprocess` header-only library

Upstream repo: https://github.com/arun11299/cpp-subprocess
Commit: 4025693decacaceb9420efedbf4967a04cb028e7

The "Convenience Functions" section is unused in our codebase, so it has
been removed.

hebasto commented:
@Sjors 

> I didn't test Windows native. It might be worth splitting off the last commit to get this merged faster. I'm assuming it's more important to get rid of boost process than it is to support native Windows builds.

I've reworked this PR to remove all Windows-specific changes. Any remaining Windows-related modifications will be addressed in follow-ups.

@theStack 

> Nit: if you have to retouch, could update the commit hash reference in the first commit body (the last upstream patch [arun11299/cpp-subprocess#98](https://github.com/arun11299/cpp-subprocess/pull/98) was merged a few weeks ago). Regarding the TODO in the same message, are there any more changes included here which would make sense to upstream? Comparing cpp-subprocess master branch with the version included here still shows some small differences:

I've reworked this PR to remove all Windows-specific changes. Therefore, diffs you mentioned are not needed in this PR. However, I'll continue to work on submitting Windows-related fixes to the upstream.

@fanquake 

> Concept ACK on removing Boost Process. Is the plan to also delete any unused code from this header? I don't think copy-pasting dead/untested code into this repository is a good idea, and only serves to complicate review, and potentially cause issues in future (it maybe already doesn't work with c++20?)

I've removed the entire "Convenience Functions" section. While other removals, such as unused API overloads, are still possible, I've opted to keep them for now to facilitate the reviewing process.

---

This PR is based on #29489 now, so please review it first :)

hebasto commented on build_msvc/bitcoin_config.h.in:
```diff
@@ -41,6 +41,9 @@
 /* Define this symbol to enable ZMQ functions */
 #define ENABLE_ZMQ 1
 
+/* define if external signer support is enabled (requires Boost::Process) */
```
Not relevant since the recent [update](https://github.com/bitcoin/bitcoin/pull/28981#issuecomment-1968539836).

Sjors commented:
#29489 was merged, please rebase :-)

hebasto commented:
> #29489 was merged, please rebase :-)

Rebased :)

Sjors commented:
re-ACK 86e1f24f20335f9790db95f253c8f493a7f63983

Mainly just drops windows support since my last review. I briefly re-tested on Ubuntu.

theStack reviewed (approved):
Light ACK 86e1f24f20335f9790db95f253c8f493a7f63983

Didn't review any of the code in cpp-subprocess, but the library seems to do its job at least on POSIX systems (tested on OpenBSD 7.4 and Ubuntu 22.04.3 LTS).

I checked for remaining boost::process leftovers/references via `git grep -i boost.*process` and found a depends patch that can likely also be removed: https://github.com/theStack/bitcoin/commit/cf78210018ad1ea83b5b035c15949045fc69c583 (it's the only depends patch for boost)

achow101 commented:
ACK 86e1f24f20335f9790db95f253c8f493a7f63983

Briefly tested on Arch, and it works. The changes to our code are straightforward and just modifying to use the new API. I looked at cpp-subprocess itself, and there definitely is a lot of code that we don't use that could be pruned in a followup. For unix systems, it indeed does the expected fork-exec. I did not look at the Windows specific code as that is currently unused.

@TheCharlatan @fanquake @dergoegge Since you Concept ACK'd, would you mind also giving this a review?

hebasto commented:
Rebased due to the conflict with the merged #29479.

@Sjors @theStack @achow101 sorry for invalidating your acks :)

achow101 commented:
reACK d5a715536e497c160a2520f81334aab6c7490213

Checked diff is just for resolving rebase conflicts.

theStack reviewed (approved):
Light re-ACK d5a715536e497c160a2520f81334aab6c7490213

(as per `git range-diff 86e1f24...d5a7155`)

Sjors commented:
re-tACK d5a715536e497c160a2520f81334aab6c7490213

Tested display address from the GUI on a Trezor on Ubuntu.

fanquake reviewed (approved):
ACK d5a715536e497c160a2520f81334aab6c7490213 - with the expectation that this code is going to be maintained as our own. Next PRs should:
* Remove linter exclusions and fix all issues.
* Delete all unused code.
* Rewrite in C++20.
* hpp -> cpp ?

hebasto commented:
> ... with the expectation that this code is going to be maintained as our own. Next PRs should:
> 
>    * Remove linter exclusions and fix all issues.

See:
- https://github.com/bitcoin/bitcoin/pull/29849
- https://github.com/bitcoin/bitcoin/pull/29910

>    * Delete all unused code.

The work has been started in https://github.com/bitcoin/bitcoin/pull/29865.
